### DISCUSSION TITLE
CMake Generators for "Quick Start a Hello World project (C++, CMake)"

### DISCUSSION THREAD

**Discussion by nicreuss** [xpack/vscode-xpack-extension-ts#30] [Q&A]
I was attempting to use an alternative generator with CMake in the quick start project. After installing `@xpack-dev-tools/windows-build-tools` in the "Quick Start a Hello World project (C++, CMake)" the `make` utility is installed in the `xpacks/.bin` directory.

Changing `Ninja` to `Unix Makefiles`  in project.json gives
```json
"commandPrepare": "cmake -S meta -B {{ properties.buildFolderRelativePath }} -G \"Unix Makefiles\" -D CMAKE_BUILD_TYPE={{ properties.buildType }} -D CMAKE_EXPORT_COMPILE_COMMANDS=ON",
```

Running the prepare task then results in an error:
```shell
cmake -B build\debug -G "Unix Makefiles" -D CMAKE_BUILD_TYPE=Debug -D CMAKE_EXPORT_COMPILE_COMMANDS=ON -D CMAKE_TOOLCHAIN_FILE=arm-gcc-toolchain.cmake

CMake Error: Could not create named generator "Unix
```

A variation of this error occurs for any quoted generator. For example, using `-G \"Ninja\"` also results in the error:
```shell
CMake Error: Could not create named generator "Ninja"
```

Is there something that I am missing that is causing this issue? Are there any other ways to use multi-word CMake generators without enclosing them in quotes?

**Comment by ilg-ul** [comment]
I'm not sure what causes the error, it seems to be a problem with the inner quotes. 

As a workaround, could you try apostrophes? In my environment (macOS) apostrophes worked, but I did not try on Windows.

**Reply by nicreuss** [reply]
Unfortunately apostrophes (like `-G 'Ninja'` or `-G 'Unix Makefiles'`) lead to the same error.

**Reply by ilg-ul** [reply]
Hmmm... then it must be an interaction with the cmd.exe used to start applications on Windows. :-(

**Reply by nicreuss** [reply]
Windows does seem to be problematic with the shims/links in the xpack system. How difficult would it be for xpm to copy over the binaries into each project instead of using shims/links (resulting in a bloated folder like node_modules)?

**Reply by nicreuss** [reply]
I have found a bizarre workaround for Windows. After trying all sorts of escape patterns for double quotes within a liquidjs template string  (`\"`, `\\\"` ... etc) I pasted in the double prime symbol `″` and the command now runs correctly.
```json
"commandFlash": "{{ properties.commandOpenocd }} -f interface/stlink.cfg -f target/stm32wbx.cfg -c ″{{ properties.argFlash }}″",
```

**Comment by ilg-ul** [comment]
I confirm that the problem is Windows specific, on macOS both quotes and apostrophes work.

I'll try to run some tests on Windows, I don't know yet where the problem is.

**Comment by ilg-ul** [comment]
> How difficult would it be for xpm to copy over the binaries into each project instead of using shims/links

That wouldn't be a reasonable solutions, some binary xPacks, like toolchains, are very large (hundreds of MB), and since dependencies can be installed for each build configuration, you might end up with a lot of space wasted.

Only shims on Windows are problematic, symbolic links are fine on Linux/macOS.

And, BTW, shims were introduced by npm, xpm simply used the same implementation.

**Comment by ilg-ul** [comment]
The alternate solution is to use Windows links, but the wise guys at Microsoft decided to make them available only after you enable the developer mode. Would you do this on your machine?

**Reply by nicreuss** [reply]
I had to enable developer mode in the past (I think it was something to do with working with Virtual Machines). So yes I think requesting windows users to enable developer mode is worth the benefits of using links.

**Comment by ilg-ul** [comment]
Another possible workaround would be to completely remove the -G, since the `Unix Makefiles` value is the default.

**Reply by nicreuss** [reply]
Unfortunately the default on Windows is either Visual Studio or NMake so Unix Makefiles isn't available without being explicitly set.

**Reply by ilg-ul** [reply]
That's unfortunate, since I don't have another workaround :-(

**Comment by ilg-ul** [comment]
But, just out of curiosity, what problems did you encounter with Ninja to make you want to go back to Make? As far as I can tell, I had no problems at all. And build systems like Meson do not even support Make by design.

**Reply by nicreuss** [reply]
I am converting a Makefile generated by STM32CubeMX to CMakeLists.txt. I wanted to compare the Makefiles generated by CMake to the original Makefiles. After encountering the behavior described above, I checked if it existed in the quick start project too.

**Comment by ilg-ul** [comment]
> I wanted to compare the Makefiles generated by CMake to the original Makefiles

Ah, for a simple test you ca simply set the path to include xpacks/.bin and run the cmake line by hand. But don't expect miracles, the code generated by automated tools is generally hard to read, and includes lots of apparently useless sequences.

BTW, I would not make any changes to the code generated by CubeMX, since by definition it is intended to be re-generated at any time, so unless you can automate the changes with a script, doing the same conversion by hand again and again is not realistic.

**Reply by nicreuss** [reply]
Very true!

Only the source files, defines and includes in the Makefile vary as outputs of CubeMX so a quick copy paste into the CMakeLists.txt solves that issue. I've had to manually make this conversion since the Makefile requires .cmd extensions for the xpack binaries on Windows. Since I would have had to modify the Makefile to detect windows, it made sense to move across to CMake instead.

I'd be open to any suggestion or better ideas - still a bit of a novice at this.

**Comment by ilg-ul** [comment]
> I've had to manually make this conversion since the Makefile requires .cmd extensions for the xpack binaries on Windows

Oops! That's a new one. 

I thought that the explicit `.cmd` extension was needed by VS Code to properly parse the `compile_commands.json`.

Aren't the projects generated by CubeMX able to run on Windows directly when using the xPack environment?

**Comment by nicreuss** [comment]
The Makefile from CubeMX can be run directly using an xpack build configuration in package.json. However, there is no intellisense support. The [Makefile vscode plugin](https://marketplace.visualstudio.com/items?itemName=ms-vscode.makefile-tools) can generate a compile_commands.json and is a configuration provider for C/C++ intellisense. However, the system headers from `arm-non-eabi-gcc` are missing in intellisense (same bug as #28). This can be rectified using a .cmd suffix in the makefile.

**Comment by ilg-ul** [comment]
Ok, so if the VS Code C/C++ extension is fixed to consider the .cmd suffix, or if I change xpm to autodetect the developer mode and create file links instead of .cmd shims, the problem should be solved, right?

For this particular case using file links might be enough, since you are using only binary xPacks, but if your build scripts would invoke npm applications, the shims would also be problematic.

So the correct solution would be to fix VS Code, but until then, using file links would be an efficient workaround for most cases.

**Reply by nicreuss** [reply]
I think autodetecting developer mode and using file system links would be a great addition to xpacks. It could also help solve other windows issues

**Comment by ilg-ul** [comment]
https://github.com/xpack/xpm-js/issues/127

**Comment by ilg-ul** [comment]
I did some experiments with file links. Creating them was easy, and, after enabling the Developer Mode, xpm automatically used this mode.

Unfortunately it looks like there is a good reason why these links are not available by default. And this reason is that... the soft file links are not recognised by legacy Windows components, including cmd.exe, used to spawn processes.

So... not very good news :-(

**Reply by nicreuss** [reply]
This might be a silly question, but is it possible to use PowerShell for launching processes instead of cmd?

**Comment by ilg-ul** [comment]
It is not a silly question, it is possible to use PowerShell, just that, for historical reasons, everybody uses cmd.exe. I did not encounter a single application to use PowerShell, and Microsoft, even at this date, exports an environment variable which is intended to be used as default shell for launching sub-processes, and this variable points to cmd.exe. I can try to redefine it, but, given the huge mess that Windows design is, I expect something else to break. 

It looks like on Windows the extensions must be explicit (.cmd, or .ps1) and applications must be aware of them.

For completeness, I will add .ps1 shims, as npm does, but I doubt that it'll help much.

In this case the problem is the VS Code extension, I have to check it to better understand if we have another solution.

**Comment by ilg-ul** [comment]
I did a more detailed test with file symlinks, and the reason why it fails is the way paths are processed.

In my test I used a link to ninja.exe and cmd.exe is able to start an executable behind a symlink, but, in this case, ninja.exe fails because it cannot load its dependency (libwinpthread-1.dll), which is in the same folder as the original file, but is missing in the location where the link is.

On non-Windows platforms there is a RUNPATH/RPATH mechanism to point the loader to the actual location of dependencies, but on Windows things are different, and DLLs are usually located in the same folder as the executable. This does not play well with the way symlinks were implemented by Microsoft, since their loader does not search the actual folder, it searches only the folder where the link is.

I would say that Windows file symbolic links are a dead end (like many of the Windows internals).

**Comment by ilg-ul** [comment]
In [xpm@0.12.1](https://xpack.github.io/blog/2021/12/30/xpm-v0-12-1-released/) I added support for PowerShell shims.

To be correct, I did nothing special, I just updated the dependencies to the latest npm module doing this.

Now, in addition to xxx.cmd, you can also use xxx.ps1 to invoke applications. I doubt the VS Code IntelliSense extension will be happier, but it is worth giving it a try.

**Reply by nicreuss** [reply]
So I've done some testing with v0.12.1 on Windows and this is what I found:
- As predicted, IntelliSense still does not work without an explicit `.cmd` extension for the compiler. In fact, adding the `.ps1` extension gives the same result as not having an extension at all.
- The Makefile tools extension gives an error trying to create a process running the ps1 file.
```powershell
process_begin: CreateProcess(NULL, arm-none-eabi-gcc.cmd -c -mcpu=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=hard -DUSE_HAL_DRIVER -DSTM32WB55xx -DDEBUG -ICore/Inc -IDrivers/STM32WBxx_HAL_Driver/Inc -IDrivers/STM32WBxx_HAL_Driver/Inc/Legacy -IDrivers/CMSIS/Device/ST/STM32WBxx/Include -IDrivers/CMSIS/Include -Og -Wall -fdata-sections -ffunction-sections -g -gdwarf-2 -MMD -MP -MFbuild/main.d -Wa,-a,-ad,-alms=build/main.lst Core/Src/main.c -o build/main.o, ...) failed.
make (e=2): The system cannot find the file specified.

make: *** [Makefile:180: build/main.o] Error 2
```
- The xPack Actions build task for my makefile project fails ( I think because its now run by powershell)
```powershell
> Executing task: xpm run build --config Debug <
make all DEBUG=1 BUILD_DIR=build/debug
mkdir -p build/debug
The syntax of the command is incorrect.
make: *** [Makefile:196: build/debug] Error 1
> exit(2)
The terminal process "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -Command xpm run build --config Debug" terminated with exit code: 1.
  ```


If I revert back to v0.11.2 then all these errors go away.

**Comment by ilg-ul** [comment]
Hmmm... Not good... :-(

> The Makefile tools extension gives an error trying to create a process running the ps1 file.
> process_begin: CreateProcess(NULL, arm-none-eabi-gcc.cmd -c 

Can you clarify exactly what this means? You mention .ps1, but the error refers to .cmd.

> The xPack Actions build task for my makefile project fails

I need to better understand if there is something wrong in xpm, to fix it. 

In my test environment, I created the hello world project, with CMake, and the build passed. I think my setup is common, I don't remember doing anything special. 

Could you try the same on your machine, with 0.12.1?

**Reply by nicreuss** [reply]
## Environment
- Windows 10 (developer mode on)
- xpack vs code extension v0.4.16
- xpm v0.12.1

## CMake Hello World: Evaluating powershell shims
Changing `.cmd` to `.ps1` in `toolchain-gcc.cmake`:

```cmake
# TODO: remove workaround once VS Code is fixed.
# VS Code does not properly identify the shims used by npm/xpm,
# thus make the extension explicit.
if("${CMAKE_HOST_SYSTEM_NAME}" STREQUAL "Windows")
  set(CMAKE_C_COMPILER "${CMAKE_C_COMPILER}.ps1")
  set(CMAKE_CXX_COMPILER "${CMAKE_CXX_COMPILER}.ps1")
endif()
```
Output:
```shell
> Executing task: xpm run prepare-all <

> xpm run prepare --config Debug
> cmake -S meta -B build\debug -G Ninja -D CMAKE_BUILD_TYPE=Debug -D CMAKE_EXPORT_COMPILE_COMMANDS=ON -D CMAKE_TOOLCHAIN_FILE=toolchain-gcc.cmake
-- CMake version: 3.20.6-gc90b991
-- The C compiler identification is unknown
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: C:/Users/nic/Desktop/xpack-test/xpacks/.bin/gcc.ps1
-- Check for working C compiler: C:/Users/nic/Desktop/xpack-test/xpacks/.bin/gcc.ps1 - broken
CMake Error at C:/Users/nic/AppData/Roaming/xPacks/@xpack-dev-tools/cmake/3.20.6-2.1/.content/share/cmake-3.20/Modules/CMakeTestCCompiler.cmake:66 (message):
  The C compiler

    "C:/Users/nic/Desktop/xpack-test/xpacks/.bin/gcc.ps1"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: C:/Users/nic/Desktop/xpack-test/build/debug/CMakeFiles/CMakeTmp

    Run Build Command(s):C:/Users/nic/Desktop/xpack-test/xpacks/.bin/ninja cmTC_81b0f && [1/2] Building C object CMakeFiles\cmTC_81b0f.dir\testCCompiler.c.obj
    FAILED: CMakeFiles/cmTC_81b0f.dir/testCCompiler.c.obj
    C:\Users\nic\Desktop\xpack-test\xpacks\.bin\gcc.ps1    -o CMakeFiles\cmTC_81b0f.dir\testCCompiler.c.obj -c testCCompiler.c
    gcc.exe: fatal error: no input files
    compilation terminated.
    ninja: build stopped: subcommand failed.

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:22 (enable_language)

-- Configuring incomplete, errors occurred!
See also "C:/Users/nic/Desktop/xpack-test/build/debug/CMakeFiles/CMakeOutput.log".
See also "C:/Users/nic/Desktop/xpack-test/build/debug/CMakeFiles/CMakeError.log".
> exit(1)
> exit(1)
The terminal process "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -Command xpm run prepare-all" terminated with exit code: 1.
```
## Personal Makefile project: testing xpm v0.12.1
### .ps1
> Can you clarify exactly what this means? You mention .ps1, but the error refers to .cmd.

Sorry this was a mistake on my part. Here is the correct terminal output from the Makefile extension with the `makePath` set to `"xpacks/.bin/make.ps1"`

```shell
Generating configuration cache with command: 'xpacks\.bin\make.ps1 all GCC_PATH=xpacks/.bin DEBUG=1 --dry-run --always-make --keep-going --print-directory'
Writing the dry-run output: c:\Users\nic\Desktop\sundial\sundial-embedded\reader\reader-rev1\.vscode\makefile\dryrun.log
Generating dry-run elapsed time: 0.588
The make dry-run command failed.
IntelliSense may work only partially or not at all.
process_begin: CreateProcess(NULL, arm-none-eabi-gcc.ps1 -c -mcpu=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=hard -DUSE_HAL_DRIVER -DSTM32WB55xx -DDEBUG -ICore/Inc -IDrivers/STM32WBxx_HAL_Driver/Inc -IDrivers/STM32WBxx_HAL_Driver/Inc/Legacy -IDrivers/CMSIS/Device/ST/STM32WBxx/Include -IDrivers/CMSIS/Include -Og -Wall -fdata-sections -ffunction-sections -g -gdwarf-2 -MMD -MP -MFbuild/main.d -Wa,-a,-ad,-alms=build/main.lst Core/Src/main.c -o build/main.o, ...) failed.
make (e=2): The system cannot find the file specified.

make: *** [Makefile:180: build/main.o] Error 2
```

Running the xPack build action gives:
```shell
> Executing task: xpm run build --config Debug <

> make all DEBUG=1 BUILD_DIR=build/debug
mkdir -p build/debug
The syntax of the command is incorrect.
make: *** [Makefile:196: build/debug] Error 1
> exit(2)
The terminal process "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -Command xpm run build --config Debug" terminated with exit code: 1.
```

### .cmd
The Makefile extension (truncated) output when using `.cmd` for comparison.
```shell
Generating configuration cache with command: 'xpacks\.bin\make.cmd all GCC_PATH=xpacks/.bin DEBUG=1 --dry-run --always-make --keep-going --print-directory'
Writing the dry-run output: c:\Users\nic\Desktop\sundial\sundial-embedded\reader\reader-rev1\.vscode\makefile\dryrun.log
Generating dry-run elapsed time: 0.177
Preprocessing: "c:\Users\nic\Desktop\sundial\sundial-embedded\reader\reader-rev1\.vscode\makefile\dryrun.log"
Preprocess elapsed time: 0
Parsing for IntelliSense.
Updating the CppTools IntelliSense Configuration Provider.
```

Running the xPack build action gives:
```shell
> Executing task: xpm run build --config Debug <

> make all DEBUG=1 BUILD_DIR=build/debug
mkdir -p build/debug
The syntax of the command is incorrect.
make: *** [Makefile:196: build/debug] Error 1
> exit(2)
The terminal process "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -Command xpm run build --config Debug" terminated with exit code: 1.
```

Notice how the build fails despite the the use of cmd. For v0.11.2 this build was successful.

## No extension
Makefile extension output:
```shell
Generating configuration cache with command: 'xpacks\.bin\make.exe all GCC_PATH=xpacks/.bin DEBUG=1 --dry-run --always-make --keep-going --print-directory'
Writing the dry-run output: c:\Users\nic\Desktop\sundial\sundial-embedded\reader\reader-rev1\.vscode\makefile\dryrun.log
Generating dry-run elapsed time: 0.04
The make dry-run command failed.
IntelliSense may work only partially or not at all.
'xpacks\.bin\make.exe' is not recognized as an internal or external command,
operable program or batch file.
```
Note the behavior of the extension is to append `.exe`.

Build task:
```shell
> Executing task: xpm run build --config Debug <

> make all DEBUG=1 BUILD_DIR=build/debug
mkdir -p build/debug
The syntax of the command is incorrect.
make: *** [Makefile:196: build/debug] Error 1
> exit(2)
The terminal process "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -Command xpm run build --config Debug" terminated with exit code: 1.
```
Oddly, this commands makes a folder called `-p` and then throws the error above. When typing `mkdir -p build/debug` into Powershell directly, it behaves as expected.

## Package.json
I have attached the [package.json](https://github.com/xpack/vscode-xpack-extension-ts/files/7794830/package.zip) for reference. The `.cmd` was changed to `.ps1` and the xpm version upgraded to v0.12.1 to perform the above tests.

**Comment by ilg-ul** [comment]
Oh, too many details, I might miss some points. From the end up:

- I like that you use substitutions in `package.json`, in my opinion liquidjs is a very powerful engine, and I plan to use it in all places requiring substitutions
- for simple cases, the definition of `argDebug` with an `{% if ...}` is fine, but if the number of build configurations increases, you can simply define properties `argDebug` with the desired values for each configuration, the configuration properties take precedence over project properties

The addition of the `.ps1` shim normally should not break anything by itself. You presented a case where a configuration using `.cmd` worked with the previous release, and does not with the 0.12.1, which worries me; we need to better understand if the problem is in xpm or in the use case.

All other tests with replacing `.cmd` by `.ps1` are nice, but frankly I have serious doubts that they are realistic. Actually I have serious doubts that using PowerShell is a good idea.

The problem is that everywhere else is using `cmd.exe`, which is happy only with `.cmd` shims. All programs that I pack as binary xPacks (CMake, ninja, make, sh), together with npm/xpm, they all use `cmd.exe` to spawn child processes, and this was the common practice on Windows for years.

Normally they should check a sequence of cases:
- if the program is an .exe, start it directly, possibly after following the symlink
- if the program is an .ps1, start it with powershell /Command "..."
- if the program is an .cmd, start it with cmd.exe /c "..."
- otherwise use the  `comspec` environment variable (default to cmd.exe) to start anything else (like .bat or any other weird Windows things)

I can implement this in xpm, since I have control over the entire sources, and I might try to suggest such patches to do this in ninja and CMake, but I'm not sure they'll be accepted upstream. Until then, the `.ps1` shims will probably not be very useful, at least in the xPack environment I have now. Then there is also GNU `make.exe` and the BusyBox `sh.exe`, which also need patches.

So, for the moment, there is not much that we can do to fix a mess created by Microsoft in decades of poor design decisions.

Thus I suggest we focus on clarifying if the latest xpm is ok or not. For my use case, if it can build the projects generated by the hello-world template, it is ok.

I would appreciate if you could use the latest hello-world template (0.5.2) and instantiate the 3 types of projects (CMake, meson, make) and let me know if the build passed.

Please do this by hand, in a `cmd.exe` terminal, in 3 empty folders, as described in the [README](https://github.com/xpack/hello-world-template-xpack), and let me know the result. Then repeat the test in 3 different folders, using a PowerShell terminal.

If the first 3 tests pass, then from my point of view xpm is ok. If the tests also run from a PowerShell, this is a bonus, but I do not guarantee anything.

Then, if you manage to patch some of the files to help the C/C++ Intellisense, and suggest improvements to the template, I can apply them.

**Comment by ilg-ul** [comment]
I also encountered an error with the new xpm. :-(

For now I reverted xpm@latest to 0.11.2, and only xpm@next is 0.12.1.

The error I see is:

```console
> cmake -S . -B build/native-cmake-debug -G Ninja -D CMAKE_BUILD_TYPE=Debug -D PLATFORM_NAME=native -D CMAKE_EXPORT_COMPILE_COMMANDS=ON
The filename, directory name, or volume label syntax is incorrect.
> exit(1)
```

I have no idea yet what this means. It happens only on GitHub Actions; the same configuration passes on my Windows 11 test machine. It is not related to the -B path, since I tried both forward and backward slashes and they both fail.

I'll further investigate.
